### ISSUE TITLE
Combobox Accessibility Issues

### ISSUE THREAD

**Issue by BryceStevenWilley** [SuffolkLITLab/docassemble-AssemblyLine#548]
Came across a use of a combo box in ILAO's Fee Waiver interview when doing an accessibility audit. There are a few accessibility issues I found when using it with a screen reader and keyboard:

* There is a button on the right side of the field that doesn't have any labels. Screen readers will announce it as just "Push button" or "button". I think the right way to describe it would be as a collapsible button, but...
* when you use keyboard controls to press the button, it quickly shows the available items, but then quickly selects the top option. What's actually happening is that on key down, the available items are shown, and then you can use the arrow keys to navigate and choose the right option, and on key up, that option is selected and entered into the combobox input field. This definitely isn't standard keyboard controls for a button like that (having to hold down one key while using others to choose something), and should be changed to be on 2 different key presses.
  * after writing this, I discovered that this only seems to happen on Firefox on Ubuntu, and not Firefox on windows, so I'm not sure how to start debugging OS differences here.
* the options available in the combobox aren't read out on the screen reader. With Orca (Ubuntu), nothing is read at all, and with NVDA (windows), it says "blank" each time you navigate to the next year selection.
* If you enter an invalid number and navigate to continue, the validation message pops up, but nothing is read out on with the reader (at all). If you navigate back to the combo box, it won't read out the issue either. If you just didn't enter anything it will read out "this field is required".

The combobox is a [separate bootstrap plugin](https://github.com/danielfarrell/bootstrap-combobox) that we would have to patch separately from docassemble. There haven't been any substantive merged PRs since 2019, so our best bet is to patch it ourselves, see if the author has the time to review and merge, and if not, convince Jonathan to use our branch instead.

Here's an image of what the combobox looks like, and some example code to try it out.

![Screenshot from 2022-08-05 15-22-45](https://user-images.githubusercontent.com/6252212/183146867-0563d49a-ad45-447f-88c4-052ff87657ff.png)

```yaml
question: |
  What year were you born?
fields:
  - Year of birth: birth_year
    input type: combobox
    datatype: integer
    maxlength: 4
    min: ${ today().year-120 }
    max: ${ today().year }
    code: |
      reversed(range(today().year-120, today().year+1))
```

**Comment by BryceStevenWilley** [comment]
Since this is (IMO) a pretty big accessibility issue, and there are a number of things to patch (that I would have to figure out, given I'm not a front-end expert yet), I might recommend that our interviews avoid combo box selections (at least the e-filing ones).

**Comment by BryceStevenWilley** [comment]
From our conversation at the coding deep-dive: 
* the particular use case of combo boxes should be to always let a user add their own option, even if not present in the drop down.
  * the other alternate way of doing this in a docassemble interview is a dropdown with the "other" choice, and a show if that let's you add another
* it'll be a complicated change either way, so we will have to defer a sweeping audit / change of existing interviews that use this feature, likely until we add e-filing features to them.

Our priorities here should be to:
* [ ] find a better plugin if possible, one that was built with accessibility in mind from the get go
* [ ] research the alternatives to using a combo box
* [ ] As a stopgap, in the [existing interviews that do use combo boxes](https://github.com/search?q=org%3ASuffolkLITLab+combobox&type=code), we should note that they're deprecated and add a link to this interview and to the documentation page that describes why we're suggesting avoiding it for now.

**Comment by BryceStevenWilley** [comment]
Did a little more searching for other plugins, and unfortunately the only one I found was https://github.com/cchantep/bootstrap-combobox, which isn't really a combobox, it's just a dropdown. There are a lot of dropdown plugins, but not a lot of combo boxes.

It's not a plugin, but w3 does have [a (not-production ready) version](https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-autocomplete-list.html), which looks good, but likely won't be a drop-in-replacement when trying to get the styles we want.

**Comment by BryceStevenWilley** [comment]
Some quick updates here: I forgot that docassemble already it's own fork of most things, including the bootstrap-combobox plugin. So that makes it easier to merge into docassemble more quickly (might also merge upstream if I have the time).

Realizing this I was able to fairly quickly make the combobox much more accessible, with the right aria properties and roles. The code is fairly done, and [in this commit](https://github.com/BryceStevenWilley/docassemble/commit/98d1896348451148219c76a8c50278a2d0214b60). I have an fairly extensive testing plan though, because I want to do this right on mobile and different operating systems, and am kinda doing it on my free time to learn more about front-end stuff.

Keeping track of the testing plan here, cause I might as well.

## OS/platforms
* [ ] ubuntu
  * [ ] Firefox
  * [ ] Chrome
  * [ ] Opera (?)
* [ ] Android
  * [ ] Chrome
  * [ ] Firefox
* [ ] Windows
  * [ ] Firefox
  * [ ] Chrome
  * [ ] Edge
* [ ] Mac
  * [ ] Safari
  * [ ] Chrome
  * [ ] Firefox
* [ ] iOS
  * [ ] Safari
  * [ ] Chrome (though IIRC, all browsers on iOS are the same engine)      

## Behaviors to test

* [ ] only keyboard controls
  * [ ] can tab into the widget
  * [ ] down and up to select existing options
  * [ ] enter or tab to select it
  * [ ] escape closes the dropbox
  * [ ] screen reader works on each action
* [ ] only mouse
  * [ ] click to select (test box or button should expand)
  * [ ] scroll / selection works
  * [ ] toggle with just the button
  * [ ] x button clears it
  * [ ] screen reader
* [ ] typing
  * [ ] can focus on the text input with a click / tap
  * [ ] can type and filter data shows
  * [ ] type partial and arrow keys down / tap to select
  * [ ] screen reader works and reads normally

**Comment by BryceStevenWilley** [comment]
https://github.com/jhpyle/docassemble/pull/581 was merged upstream and released as 1.4.13. However there are a few remaining things to address here:
* the known bugs:
  * [ ] Android + Firefox + Talkback doesn't handle focus well to chose from the filtered list
  * [ ] Mac + Safari + VoiceOver just doesn't work / read out the options. Need to learn more about talk back, and figure out what is supported and what's cause the specific issue.
* [x] still needs testing on iPhone. Not optimistic with how VoiceOver behaved on Mac though
* [x] The issue with validation messages not being read (copied below) hasn't been addressed. Even though I'd say the example given was a bit non-standard, validation code on comboboxes is still possible, so should be addressed at some point. 

> If you enter an invalid number and navigate to continue, the validation message pops up, but nothing is read out on with the reader (at all). If you navigate back to the combo box, it won't read out the issue either. If you just didn't enter anything it will read out "this field is required".

**Comment by nonprofittechy** [comment]
@BryceStevenWilley can we get an update on upstream status. I've seen at least one more changelog entry dealing with combobox. Is it still not fully accessible?

**Comment by BryceStevenWilley** [comment]
None of those changes have dealt with the known issues listed in https://github.com/SuffolkLITLab/docassemble-AssemblyLine/issues/548#issuecomment-1292238147. Mac / VoiceOver support is the biggest one remaining; I'd be willing to mark this as closed if we could get that working. I haven't had to time to dive into VoiceOver / Safari compatibility stuff though.

**Comment by BryceStevenWilley** [comment]
Revisiting this on Mac+Safari and iPhone, and the situation is better, but it still needs a bit of work.

> Android + Firefox + Talkback doesn't handle focus well to chose from the filtered list

I no longer have a working Android phone to test with here, I can take a bit more time to set up an emulator soon.

> Mac + Safari + VoiceOver just doesn't work / read out the options

I'm not sure exactly what I meant here, but VoiceOver doesn't just read out options for other browser builtin inputs (like selects) either. I would consider this broken though, because VoiceOver specifically says to "type Control-Option Space to see the list of options", and that does not work for this implementation. Looks like it's because that shortcut does a "click" event into the combobox, and ours doesn't react to that. Given that the [W3C example of a combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-list/) does show the combobox options with a "click", we should as well. Working on that now.

> still needs testing on iPhone.

iPhone has the same issues as Mac Voiceover, i.e. just the "click" event needs to show the combobox options.

> The issue with validation messages not being read (copied below) hasn't been addressed.

That seems to work okay now. Per-field validation is read when the user tries to continue (same time that it shows up the screen), and is read when navigating back to the item (it's treated as text around the item, so not directly read when focus is on the input, but I don't know if that's standard) (other types of validation in docassemble are bad though, I might make another issue for that).

**Comment by nonprofittechy** [comment]
Checking in Firefox for Android on Pixel 7 and I can't interact with the drop-down menu. It still works fine in Chrome.

edit: FWIW, I think Firefox is very unpopular on Android, so I don't know if we need to plan around true bugs in Firefox, as long as we're standards compliant in a reasonable way.

**Comment by BryceStevenWilley** [comment]
Have to set this down for the moment. Current progress is at https://github.com/SuffolkLITLab/docassemble/tree/combobox_accessibility
